documentation/modules/oauth/con-oauth-authorization-keycloak-example.adoc

```diff
@@ -0,0 +1,481 @@
+  --producer.config /tmp/team-b-client.properties
+Message 1
+```
+
+We get a `Not authorized to access topics: [x_messages]` error as we expected.
+Client `team-a-client`, on the other hand, should be able to write to such a topic:
+
+```
+bin/kafka-console-producer.sh --broker-list my-cluster-kafka-bootstrap:9092 --topic x_messages \
+  --producer.config /tmp/team-a-client.properties
+Message 1
+```
+
+However, we again receive `Not authorized to access topics: [x_messages]`. What's going on?
+The reason for failure is that while `team-a-client` can write to `x_messages` topic, it does not have a permission to create a topic if it does not yet exist.
+
+We now need a power user that can create a topic with all the proper settings - like the right number of partitions and replicas.
+
+
+### Power User Can Do Anything
+
+Let's create a configuration for user `bob` who has full ability to manage everything on Kafka cluster `my-cluster`.
+
+We'll need some helper scripts so we can authenticate to the `keycloak` instance.
+
+Download the following `oauth.sh` and `jwt.sh` tools to `/tmp` dir and make them executable:
+
+   curl https://raw.githubusercontent.com/strimzi/strimzi-kafka-oauth/master/examples/docker/kafka-oauth-strimzi/kafka/oauth.sh -s > /tmp/oauth.sh
+   chmod +x /tmp/oauth.sh
+
+   curl https://raw.githubusercontent.com/strimzi/strimzi-kafka-oauth/master/examples/docker/kafka-oauth-strimzi/kafka/jwt.sh -s > /tmp/jwt.sh
+   chmod +x /tmp/jwt.sh
+
+Now, `bob` will authenticate to Keycloak server with his username and password and get a refresh token.
+
+   export TOKEN_ENDPOINT=http://keycloak:8080/auth/realms/kafka-authz/protocol/openid-connect/token
+   REFRESH_TOKEN=$(/tmp/oauth.sh -q bob)
+
+This will prompt you for a password. Type 'bob-password'.
+
+We can inspect the refresh token:
+
+   /tmp/jwt.sh $REFRESH_TOKEN
+
+By default this is a long-lived refresh token that does not expire.
+
+Now we will create the configuration file for `bob`:
+
+```
+cat > /tmp/bob.properties << EOF
+security.protocol=SASL_PLAINTEXT
+sasl.mechanism=OAUTHBEARER
+sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required \
+  oauth.refresh.token="$REFRESH_TOKEN" \
+  oauth.client.id="kafka-cli" \
+  oauth.token.endpoint.uri="http://keycloak:8080/auth/realms/kafka-authz/protocol/openid-connect/token" ;
+sasl.login.callback.handler.class=io.strimzi.kafka.oauth.client.JaasClientOauthLoginCallbackHandler
+EOF
+```
+
+Note that we use the `kafka-cli` public client for the `oauth.client.id` in the `sasl.jaas.config`.
+Since that is a public client it does not require any secret.
+We can use this because we authenticate with a token directly (in this case a refresh token is used to request an access token behind the scenes which is then sent to Kafka broker for authentication, and we already did the authentication for the refresh token).
```

Comment:
```suggestion
The `kafka-cli` public client is used for the `oauth.client.id` in the `sasl.jaas.config`.
Since that is a public client it does not require a Secret.
We can use this because we authenticate with a token directly. In this case, the refresh token requests an access token behind the scenes, which is then sent to the Kafka broker for authentication. The refresh token has already been authenticated.
```